module/VuFind/src/VuFind/ILS/Driver/XCNCIP2.php

```diff
@@ -1189,51 +1237,88 @@ public function getMyHolds($patron)
-        ];
-        $request = $this->getLookupUserRequest(
-            $patron['cat_username'], $patron['cat_password'],
-            $patron['patronAgencyId'], $extras
+        $response = $this->getLookupUserResponse(
+            $patron['cat_username'], $patron['cat_password']
         );
-        $response = $this->sendRequest($request);
-        $this->checkResponseForError($response);
 
-        $first = $response->xpath(
+        $firstname = $response->xpath(
             'ns1:LookupUserResponse/ns1:UserOptionalFields/ns1:NameInformation/' .
             'ns1:PersonalNameInformation/ns1:StructuredPersonalUserName/' .
             'ns1:GivenName'
         );
-        $last = $response->xpath(
+        $lastname = $response->xpath(
             'ns1:LookupUserResponse/ns1:UserOptionalFields/ns1:NameInformation/' .
             'ns1:PersonalNameInformation/ns1:StructuredPersonalUserName/' .
             'ns1:Surname'
         );
+        if (empty($firstname) && empty($lastname)) {
+            $lastname = $response->xpath(
+                'ns1:LookupUserResponse/ns1:UserOptionalFields/' .
+                'ns1:NameInformation/ns1:PersonalNameInformation/' .
+                'ns1:UnstructuredPersonalUserName'
+            );
+        }
 
-        // TODO: distinguish between permanent and other types of addresses; look
-        // at the UnstructuredAddressType field and handle multiple options.
-        $address = $response->xpath(
+        $address1 = $response->xpath(
+            'ns1:LookupUserResponse/ns1:UserOptionalFields/' .
+            'ns1:UserAddressInformation/ns1:PhysicalAddress/' .
+            'ns1:StructuredAddress/ns1:Line1' .
+            '|' .
+            'ns1:LookupUserResponse/ns1:UserOptionalFields/' .
+            'ns1:UserAddressInformation/ns1:PhysicalAddress/' .
+            'ns1:StructuredAddress/ns1:Street'
+        );
+        $address1 = !empty($address1) ? (string)$address1[0] : null;
+        $address2 = $response->xpath(
+            'ns1:LookupUserResponse/ns1:UserOptionalFields/' .
+            'ns1:UserAddressInformation/ns1:PhysicalAddress/' .
+            'ns1:StructuredAddress/ns1:Line2' .
+            '|' .
+            'ns1:LookupUserResponse/ns1:UserOptionalFields/' .
+            'ns1:UserAddressInformation/ns1:PhysicalAddress/' .
+            'ns1:StructuredAddress/ns1:Locality'
+        );
+        $address2 = !empty($address2) ? (string)$address2[0] : null;
+        $zip = $response->xpath(
             'ns1:LookupUserResponse/ns1:UserOptionalFields/' .
             'ns1:UserAddressInformation/ns1:PhysicalAddress/' .
-            'ns1:UnstructuredAddress/ns1:UnstructuredAddressData'
+            'ns1:StructuredAddress/ns1:PostalCode'
         );
-        $address = explode("\n", trim((string)$address[0]));
+        $zip = !empty($zip) ? (string)$zip[0] : null;
+
+        if (empty($address1)) {
+            // TODO: distinguish between more formatting types; look
```

Comment:
Is this still a TODO, or have you actually done some of it here?

Reply:
Not really. There is pair element for UnstructuredAddressData called UnstructuredAddressType, which can contain almost anything which describes  the structure/formatting of data element. But it describes it in pretty vague way like "HTML", "XML", "New-line delimited text" and so on. So this comment is here for future possiblity of enhancement address parsing, but until we have a live system  using this, it is almost impossible to make concrete implementation.

Reply:
Makes sense!